Murloc7 b alex vacations bjs2 40691

alexFromAC143937 opened the pull request.

alexFromAC143937 committed:
NoTest

alexFromAC143937 committed:
BJS2-40691
Любой проект может выставлять вакансии о поиске специалистов. В будущем мы сможем расширить это в post_service — можно будет писать посты о вакансиях, нотификации прикрутим и т.д. Пока что просто создаём вакансии: можно выпустить несколько одинаковым вакансий, можно описывать в них требуемый опыт и т.д. Также на проекте обязательно должен быть человек, ответственный за вакансию.

alexFromAC143937 committed:
BJS2-40691
Устранение конфликтов

alexFromAC143937 committed:
Merge branch 'Murloc-master-stream7' into Murloc7-BAlex-vacations-BJS2-40691

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/7c39ab73-7f5d-4c00-8c24-b7b35f8b3970)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/4bb4ef7f-2520-4910-8d12-175229f34257)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/bb5b08c8-f112-43eb-ac8e-da596cb67ad9)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/12294e7f-b37a-4d2c-9b11-ba43b0870245)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/59a014d7-ab9f-453f-9d13-22abe3df4750)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/fffae966-a372-496a-bb05-e02a8027f485)

RuslanGabd commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,90 @@
+@RequestMapping(UrlUtils.MAIN_URL + UrlUtils.V1 + UrlUtils.VACANCY)
+public class VacancyController {
+    private final VacationService vacationService;
+    private final ValidationVacancies validationVacancies;
```
VacationService - ОтпускСервис ?

jibulani replied:
почему зарезолвил? не исправлено же

RuslanGabd commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,90 @@
+        validationVacancies.hasToBeNotNull(vacancyDto.createdBy(), "createdBy");
+        validationVacancies.hasToBeNotNull(vacancyDto.updatedBy(), "updatedBy");
+        validationVacancies.hasToBeNotNull(vacancyDto.count(), "count");
+
```
Либо я что-то упускаю, либо я не понимаю зачем так сложно проверять на null. Почему нельзя в Dto над полем Добавить аннотацию @NotNull

jibulani replied:
тоже не исправлено

RuslanGabd commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,90 @@
+    @DeleteMapping()
+    public VacancyDto delete(@NotNull @RequestBody VacancyDto vacancyDto) {
+        validationVacancies.hasToBeNotNull(vacancyDto.id(), "id vacancy");
+        validationVacancies.isVacancyExist(vacancyDto.id());
```
зачем нам с фронта получать целое  Dto когда нужен только id? Можно из RequestParam получать только id, а еще лучше из PathVariable:

@DeleteMapping("/{id}")
public ResponseEntity<Void> delete(@PathVariable  @Min(1)  Long id) {

jibulani replied:
согласен

RuslanGabd commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,90 @@
+
+        log.info("VacancyController.delete - id: {}", vacancyDtoResult.id());
+        return vacancyDtoResult;
+    }
```
Зачем Dto возвращаешь? то есть удалили вакансию и возвращаем Dto на сущности которой уже нет, как по мне не логично.
Здесь лучше вернуть id удаленной сущности.

RuslanGabd commented on src/main/java/faang/school/projectservice/repository/VacancyExtraRepository.java:
```diff
@@ -0,0 +1,20 @@
+@Repository
+@RequiredArgsConstructor
+public class VacancyExtraRepository {
+
```
не совсем понял за чем этот класс. Почему нельзя  VacancyRepository использовать сразу в сервисе.

alexFromAC143937 replied:
Видится, что обработка должна использоваться в реализации репазирория, так как может переиспользоваться и другими классами

jibulani replied:
другие классы по-хорошему должны работать с вакансиями через VacancyService, так что действительно избыточно выглядит

RuslanGabd commented on src/main/java/faang/school/projectservice/service/CandidateService.java:
```diff
@@ -0,0 +1,28 @@
+        } else {
+            if (candidate.getCandidateStatus() != null
+                    && candidate.getCandidateStatus().name().equals(candidateStatusNotDelete.name())) {
+                candidateRepository.deleteById(idCandidate);
```
а если одно из условии не выполниться то кандидат не удалиться и мы никак об этом не узнаем. ведь ошибка не выброситься

alexFromAC143937 replied:
В данной ситуации такая бизнес логика

RuslanGabd commented on src/main/java/faang/school/projectservice/service/CandidateService.java:
```diff
@@ -0,0 +1,28 @@
+
+    public void deleteCandidateByIdByStatus(Long idCandidate, CandidateStatus candidateStatusNotDelete) {
+        Candidate candidate = candidateRepository.findById(idCandidate).orElse(null);
+        if (candidate == null) {
```
Я бы выбрасывал ошибку вместо null тогда одна проверка отпадает, а две другие в отдельный приватный метод вынести.

alexFromAC143937 replied:
Суть в том, что в рамках бизнес логики, удалить нужно, если значения есть, если нет, то продолжиь работу далее.

RuslanGabd commented on src/main/java/faang/school/projectservice/service/VacationService.java:
```diff
@@ -0,0 +1,87 @@
+        validationVacancies.personHasNecessaryRole(vacancyDto.updatedBy(), "updatedBy", vacancyDto.projectId());
+
+        return mapperVacation.vacancyToVacancyDTo(vacancyExtraRepository.getVacancyRepository().save(mapperVacation.vacancyDToToVacancy(vacancyDto)));
+    }
```
такой return сложно читаем.

Снова вопрос почему нельзя VacancyRepository использвать сразу.
У нас на входе уже есть Dto можно и его возвращать

alexFromAC143937 replied:
Проблема в том, что если использовать сразу, то нужно еще обрабатывать ислючения каждый раз, если не найден,  а так а) переиспользование б) кода наоборот меньше)

RuslanGabd reviewed (commented):
Интересно сделано, мне пару моментов не понятно, оставил комментарии.

alexFromAC143937 committed:
40691 - fix
1) Переписал мапер. Сделал опять интерфейс. Убрал обращение к репозиторию.
2) Так же поправил тесты.

jibulani commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @PostMapping(UrlUtils.VACANCY_ID)
+    public VacancyDto getVacancyById(@NotNull @RequestBody VacancyDto vacancyDto) {
```
почему для получения вакансии по id используешь Post-запрос и передаешь VacancyDto?
достаточно передать только id в аргументы и использовать Get-запрос
и вместо VACANCY_ID нужно использовать просто ID

alexFromAC143937 replied:
Задание: Контроллер принимает и возвращает только объекты VacancyDto или коллекции этих объектов. Нужно создать класс VacancyDto с минимальным необходимым набором полей для обеспечения работы функционала.

alexFromAC143937 replied:
ID - это "/{id}"; VACANCY_ID = "/id";. А так как указано, что "принимает и возвращает только объекты VacancyDto", поэтому введена отдельная переменная

jibulani replied:
тут скорее неточность формулировки, обычно делают так, как я написал
ок, как сделал, так и сделал

jibulani commented on src/main/java/faang/school/projectservice/service/VacationService.java:
```diff
@@ -0,0 +1,87 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class VacationService {
```
VacancyService

alexFromAC143937 replied:
Done

jibulani commented on src/main/java/faang/school/projectservice/service/VacationService.java:
```diff
@@ -0,0 +1,87 @@
+    public VacancyDto saveVacation(VacancyDto vacancyDto) {
+        log.info("saveVacation, vacancyDto:{} ", vacancyDto);
+
+        validationVacancies.isProjectExist(vacancyDto.projectId());
```
обычно такие методы не в отдельный класс выносят, а просто в приватные методы текущего класса

alexFromAC143937 replied:
Думал над этим, но так как валидация может проводится как в контролере так и сервисе и исходя из нетипичности работы с ДТО, пришел к выводу, что лучше вынести валидацию в один класс, откуда брать  и использовать, где нужно

jibulani replied:
для проверки дто достаточно аннотациями нужные поля пометить (`@NotNull`, `@NotBlank` и тд)

jibulani commented on src/main/java/faang/school/projectservice/validation/ValidationVacancies.java:
```diff
@@ -0,0 +1,108 @@
+@Slf4j
+@Data
+@Component
+public class ValidationVacancies {
```
если выносить в отдельный класс, то я бы назвал его ValidationVacanciesUtil

alexFromAC143937 replied:
Думал вынести в папку утилит, но так как может понадобится валидацию и для другиг сервисов, то создал папку faang.school.projectservice.validation, поэтому не целесообразно добавлять к названию Utils

jibulani commented on src/main/java/faang/school/projectservice/validation/ValidationVacancies.java:
```diff
@@ -0,0 +1,108 @@
+    private final VacancyRepository vacancyRepository;
+    private final TeamMemberRepository teamMemberRepository;
+
+    public void hasToBeNotNull(String parameterValue, String nameFieldForMessage) {
```
я бы переименовал checkNotNull

alexFromAC143937 replied:
Done

jibulani commented on src/main/java/faang/school/projectservice/validation/ValidationVacancies.java:
```diff
@@ -0,0 +1,108 @@
+        }
+    }
+
+    public void isProjectExist(Long idProject) {
```
isProjectExist намекает, что метод должен вернуть boolean (есть или нет)
в данном случае тоже назвал бы в духе checkProjectExist

jibulani replied:
в остальных местах по аналогии

jibulani commented on src/main/resources/data-project-init.sql:
```diff
@@ -0,0 +1,55 @@
+/*
```
зачем закомменченный код?

jibulani commented on src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java:
```diff
@@ -0,0 +1,258 @@
+class VacancyControllerTest {
+    @SuppressWarnings("SpringJavaInjectionPointsAutowiringInspection")
+    @Autowired
+    private MockMvc mockMvc;
```
круто, что написал тесты с mockMvc!

jibulani commented on src/test/java/utils/VacationDataCreatorTest.java:
```diff
@@ -0,0 +1,177 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class VacationDataCreatorTest {
```
это же не класс с тестами, лучше назвать просто VacancyDataCreator

alexFromAC143937 replied:
Done

jibulani reviewed (changes requested):
хорошее покрытие тестами, молодец!

есть комментарии от меня и Руслана, нужно поправить

alexFromAC143937 committed:
40691 - fix

jibulani reviewed (approved):
принято!
